File: jscomp/others/belt_Map.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(*  Adapted by authors of ReScript without using functors          *)
(***********************************************************************)

(** specalized when key type is `int`, more efficient
    than the generic type
*)
module Int = Belt_MapInt


(** specalized when key type is `string`, more efficient
    than the generic type *)
module String = Belt_MapString

(** seprate function from data, a more verboe but slightly
    more efficient
*)
module Dict = Belt_MapDict



type ('key, 'id ) id = ('key, 'id) Belt_Id.comparable
type ('key, 'id ) cmp = ('key, 'id) Belt_Id.cmp

type ('k,'v,'id) t = {
    cmp: ('k,'id) cmp;
    data: ('k,'v, 'id) Dict.t
}





let fromArray (type k idx) data ~(id : (k,idx) id)  =
  let module M = (val id) in
  let cmp = M.cmp in
  {cmp; data = (Dict.fromArray ~cmp data)}

let remove m x  =
  let {cmp; data=odata} = m in
  let newData = Dict.remove odata x ~cmp  in
  if newData == odata then m
  else {cmp; data = newData}

let removeMany m x =
  let {cmp; data=odata} = m  in
  let newData = Dict.removeMany odata x ~cmp in
  {cmp;  data = newData}

let set m key d  =
  let cmp = m.cmp in
  {cmp; data = Dict.set ~cmp m.data key d}

let mergeMany m e =
  let cmp = m.cmp in
  {cmp; data = Dict.mergeMany ~cmp m.data e}

let updateU m key f  =
  let cmp = m.cmp in
  {cmp; data = Dict.updateU ~cmp m.data key f }
let update m key f = updateU m key (fun [@bs] a  -> f a )
let split m x =
  let cmp = m.cmp in
  let (l,r),b = Dict.split ~cmp m.data x in
  ({cmp; data=l}, {cmp;  data=r}), b

let mergeU s1 s2 f =
  let cmp = s1.cmp in
  {cmp; data=(Dict.mergeU ~cmp  (s1.data) (s2.data) f)}

let merge s1 s2 f =
  mergeU s1 s2 (fun [@bs] a b c -> f a b c)

let make (type key idx) ~(id: (key, idx) id) =
  let module M = (val id) in
  {cmp = M.cmp; data = Dict.empty}

let isEmpty map =
  Dict.isEmpty map.data

let findFirstByU m f = Dict.findFirstByU m.data f
let findFirstBy m f = findFirstByU m (fun [@bs] a b -> f a b) 
let forEachU m f = Dict.forEachU m.data f
let forEach m f = forEachU m (fun [@bs] a b -> f a b)
let reduceU m acc f = Dict.reduceU m.data acc f
let reduce m acc f = reduceU m acc (fun[@bs] a b c -> f a b c )
let everyU m f = Dict.everyU m.data f
let every m f = everyU m (fun [@bs] a b -> f a b)
let someU m f = Dict.someU m.data f
let some m f = someU m (fun[@bs] a b -> f a b)
let keepU m f =
  {cmp = m.cmp; data = (Dict.keepU m.data f)}
let keep m f = keepU m (fun [@bs] a b -> f a b)

let partitionU m p =
  let cmp = m.cmp in
  let l,r = m . data |. Dict.partitionU  p in
  {cmp; data=l}, {cmp; data=r}
let partition m p = partitionU m (fun [@bs] a b -> p a b)

let mapU m f =
  {cmp = m.cmp; data = Dict.mapU m.data f}
let map m f = mapU m (fun [@bs] a  -> f a)
let mapWithKeyU m  f =
  {cmp = m.cmp; data = Dict.mapWithKeyU m.data f}
let mapWithKey m f = mapWithKeyU m (fun [@bs] a b -> f a b)
let size map = Dict.size map.data
let toList map = Dict.toList map.data
let toArray m = Dict.toArray m.data
let keysToArray m = Dict.keysToArray m.data
let valuesToArray m = Dict.valuesToArray m.data
let minKey m = Dict.minKey m.data
let minKeyUndefined m = Dict.minKeyUndefined m.data
let maxKey m = Dict.maxKey m.data
let maxKeyUndefined m = Dict.maxKeyUndefined m.data
let minimum m = Dict.minimum m.data
let minUndefined m = Dict.minUndefined m.data
let maximum m = Dict.maximum m.data
let maxUndefined m = Dict.maxUndefined m.data

let get map x  =
  Dict.get ~cmp:map.cmp  map.data x

let getUndefined map x =
  Dict.getUndefined ~cmp:map.cmp map.data x

let getWithDefault map x def =
  Dict.getWithDefault ~cmp:map.cmp map.data x  def

let getExn map x =
  Dict.getExn ~cmp:map.cmp map.data x

let has map x =
  Dict.has ~cmp:map.cmp map.data x

let checkInvariantInternal m  =
  Dict.checkInvariantInternal m.data

let eqU m1 m2 veq =
  Dict.eqU ~kcmp:m1.cmp ~veq m1.data m2.data
let eq m1 m2 veq = eqU m1 m2 (fun[@bs] a b -> veq a b)

let cmpU m1 m2 vcmp =
  Dict.cmpU ~kcmp:m1.cmp  ~vcmp m1.data m2.data
let cmp m1 m2 vcmp = cmpU m1 m2 (fun [@bs] a b -> vcmp a b)

let getData m = m.data

let getId (type key identity) (m : (key,_,identity) t) : (key, identity) id =
  let module T = struct
    type nonrec identity = identity
    type nonrec t = key
    let cmp =  m.cmp
  end in
  (module T )

let packIdData (type key idx) ~(id : (key, idx) id) ~data  =
  let module M = (val id) in
  {cmp = M.cmp ; data}

File: lib/es6/belt_Map.js


import * as Belt_MapDict from "./belt_MapDict.js";

function fromArray(data, id) {
  let cmp = id.cmp;
  return {
    cmp: cmp,
    data: Belt_MapDict.fromArray(data, cmp)
  };
}

function remove(m, x) {
  let odata = m.data;
  let cmp = m.cmp;
  let newData = Belt_MapDict.remove(odata, x, cmp);
  if (newData === odata) {
    return m;
  } else {
    return {
      cmp: cmp,
      data: newData
    };
  }
}

function removeMany(m, x) {
  let cmp = m.cmp;
  let newData = Belt_MapDict.removeMany(m.data, x, cmp);
  return {
    cmp: cmp,
    data: newData
  };
}

function set(m, key, d) {
  let cmp = m.cmp;
  return {
    cmp: cmp,
    data: Belt_MapDict.set(m.data, key, d, cmp)
  };
}

function mergeMany(m, e) {
  let cmp = m.cmp;
  return {
    cmp: cmp,
    data: Belt_MapDict.mergeMany(m.data, e, cmp)
  };
}

function update(m, key, f) {
  let cmp = m.cmp;
  return {
    cmp: cmp,
    data: Belt_MapDict.update(m.data, key, f, cmp)
  };
}

function split(m, x) {
  let cmp = m.cmp;
  let match = Belt_MapDict.split(m.data, x, cmp);
  let match$1 = match[0];
  return [
    [
      {
        cmp: cmp,
        data: match$1[0]
      },
      {
        cmp: cmp,
        data: match$1[1]
      }
    ],
    match[1]
  ];
}

function merge(s1, s2, f) {
  let cmp = s1.cmp;
  return {
    cmp: cmp,
    data: Belt_MapDict.merge(s1.data, s2.data, f, cmp)
  };
}

function make(id) {
  return {
    cmp: id.cmp,
    data: undefined
  };
}

function isEmpty(map) {
  return Belt_MapDict.isEmpty(map.data);
}

function findFirstBy(m, f) {
  return Belt_MapDict.findFirstBy(m.data, f);
}

function forEach(m, f) {
  Belt_MapDict.forEach(m.data, f);
}

function reduce(m, acc, f) {
  return Belt_MapDict.reduce(m.data, acc, f);
}

function every(m, f) {
  return Belt_MapDict.every(m.data, f);
}

function some(m, f) {
  return Belt_MapDict.some(m.data, f);
}

function keep(m, f) {
  return {
    cmp: m.cmp,
    data: Belt_MapDict.keep(m.data, f)
  };
}

function partition(m, p) {
  let cmp = m.cmp;
  let match = Belt_MapDict.partition(m.data, p);
  return [
    {
      cmp: cmp,
      data: match[0]
    },
    {
      cmp: cmp,
      data: match[1]
    }
  ];
}

function map(m, f) {
  return {
    cmp: m.cmp,
    data: Belt_MapDict.map(m.data, f)
  };
}

function mapWithKey(m, f) {
  return {
    cmp: m.cmp,
    data: Belt_MapDict.mapWithKey(m.data, f)
  };
}

function size(map) {
  return Belt_MapDict.size(map.data);
}

function toList(map) {
  return Belt_MapDict.toList(map.data);
}

function toArray(m) {
  return Belt_MapDict.toArray(m.data);
}

function keysToArray(m) {
  return Belt_MapDict.keysToArray(m.data);
}

function valuesToArray(m) {
  return Belt_MapDict.valuesToArray(m.data);
}

function minKey(m) {
  return Belt_MapDict.minKey(m.data);
}

function minKeyUndefined(m) {
  return Belt_MapDict.minKeyUndefined(m.data);
}

function maxKey(m) {
  return Belt_MapDict.maxKey(m.data);
}

function maxKeyUndefined(m) {
  return Belt_MapDict.maxKeyUndefined(m.data);
}

function minimum(m) {
  return Belt_MapDict.minimum(m.data);
}

function minUndefined(m) {
  return Belt_MapDict.minUndefined(m.data);
}

function maximum(m) {
  return Belt_MapDict.maximum(m.data);
}

function maxUndefined(m) {
  return Belt_MapDict.maxUndefined(m.data);
}

function get(map, x) {
  return Belt_MapDict.get(map.data, x, map.cmp);
}

function getUndefined(map, x) {
  return Belt_MapDict.getUndefined(map.data, x, map.cmp);
}

function getWithDefault(map, x, def) {
  return Belt_MapDict.getWithDefault(map.data, x, def, map.cmp);
}

function getExn(map, x) {
  return Belt_MapDict.getExn(map.data, x, map.cmp);
}

function has(map, x) {
  return Belt_MapDict.has(map.data, x, map.cmp);
}

function checkInvariantInternal(m) {
  Belt_MapDict.checkInvariantInternal(m.data);
}

function eq(m1, m2, veq) {
  return Belt_MapDict.eq(m1.data, m2.data, m1.cmp, veq);
}

function cmp(m1, m2, vcmp) {
  return Belt_MapDict.cmp(m1.data, m2.data, m1.cmp, vcmp);
}

function getData(m) {
  return m.data;
}

function getId(m) {
  let cmp = m.cmp;
  return {
    cmp: cmp
  };
}

function packIdData(id, data) {
  return {
    cmp: id.cmp,
    data: data
  };
}

let Int;

let $$String;

let Dict;

let cmpU = cmp;

let eqU = eq;

let findFirstByU = findFirstBy;

let forEachU = forEach;

let reduceU = reduce;

let everyU = every;

let someU = some;

let updateU = update;

let mergeU = merge;

let keepU = keep;

let partitionU = partition;

let mapU = map;

let mapWithKeyU = mapWithKey;

export {
  Int,
  $$String,
  Dict,
  make,
  isEmpty,
  has,
  cmpU,
  cmp,
  eqU,
  eq,
  findFirstByU,
  findFirstBy,
  forEachU,
  forEach,
  reduceU,
  reduce,
  everyU,
  every,
  someU,
  some,
  size,
  toArray,
  toList,
  fromArray,
  keysToArray,
  valuesToArray,
  minKey,
  minKeyUndefined,
  maxKey,
  maxKeyUndefined,
  minimum,
  minUndefined,
  maximum,
  maxUndefined,
  get,
  getUndefined,
  getWithDefault,
  getExn,
  remove,
  removeMany,
  set,
  updateU,
  update,
  mergeMany,
  mergeU,
  merge,
  keepU,
  keep,
  partitionU,
  partition,
  split,
  mapU,
  map,
  mapWithKeyU,
  mapWithKey,
  getData,
  getId,
  packIdData,
  checkInvariantInternal,
}
/* No side effect */
